사용자 API 작성

riverandeye opened the pull request:
사용자 정보에 대한 CRUD를 작성하였습니다. 

server/common에, 단독으로 id만 param으로 제공되는 경우 Validation 하는 pipe를 추가하여 중복되는 곳에 사용하였습니다. 

사용자 입력에 대한 Validation 은 아직 작성하지 못했습니다. (어떻게 해야할지 고민중) 이건 다음 PR때 기능 추가해서 구현하겠습니다. 서비스 단에서 예외처리 기능은 우선 테스트를 위해 비워놨고, 어디서 해당 예외처리를 해야 할 지 고민을 해야할 것 같습니다. (잘모르겠음)

riverandeye committed:
[server] | feat(user): User dto 정의

User dto 및 UserResponseDto 를 정의하였습니다.

riverandeye committed:
[server] | refactor(user): Get 메소드 Param 변수명 변경

userId는 User가 아닌 다른 테이블이 가지는 Foreign key 값이기 때문에, 변수명으로 부적절하다고 생각하여 변경하였습니다.

riverandeye committed:
[server] | feat(user): User api 추가

User api를 추가하였습니다.
Post /users = 사용자 추가
Delete /users/:id = id에 해당하는 사용자 제거
Put /users/:id = id에 해당하는 사용자 정보 변경

riverandeye committed:
[server] | feat(pipes): Param에 단독으로 제공된 ID를 Validate하는 pipe 작성

User 뿐만 아니라, 각 Entity의 단일 항목을 요청할 떄 반복적으로 사용되는 pipe를 작성하였습니다.
Param으로 입력된 id에 대하여 해당 id가 Validate하는 기능을 수행합니다.

riverandeye committed:
[server] | feat(user): 사용자 정보를 Validate 하는 Pipe 작성

현재 미구현된 상태이고, Validate하는 조건들을 추후에 추가할 예정입니다.

riverandeye committed:
[server] | feat(user): user api에 필요한 dto 정의

user.dto: 사용자 생성 및 수정에 필요한 dto
user-response.dto: api 결과로 리턴되는 dto에 대한 정의
user-id-pararm.dto: url로 전달되는 param dto에 대한 정의

riverandeye committed:
[server] | feat(user): pipe를 이용한 user api 정의

각 api에 Pipe를 적용하여 입력을 validate 하였습니다.
ValidateUserPipe는 현재 완성하진 못했고, 추후 업데이트해야 합니다.

hyokualexkwon commented on packages/server/src/user/pipes/validate-user.pipe/index.ts:
```diff
@@ -0,0 +1,9 @@
+
+@Injectable()
+export class ValidateUserPipe implements PipeTransform {
+  public async transform(updateUserPipeDto: ValidateUserPipeDto) {
```
본 pipe는 아무런 validation을 하는것 같지 않은데, 왜 작성하신건가요?

riverandeye replied:
어떻게 Validation을 해야 할지 고민중이여서 추후에 업데이트해서 PR 해야합니다. 형식만 만들어 놓은 것입니다.

Atercatus replied:
사용하지 않는 것을 미리 만들지 말아주세요

hyokualexkwon commented on packages/server/src/common/pipes/validate-id-param.pipe/index.ts:
```diff
@@ -0,0 +1,24 @@
+    return { id: parseInt(id, 10) };
+  }
+
+  private validateIdPipeDto({ id }: ValidateIdParamPipeDto) {
```
본 메소드에서 검사하는 로직을 하단의 validateId 에서 한번에 하면 어떨까요? 여기서는 id의 존재 여부를 검사하는 로직이 추가 된것 같은데, 별도로 분리할 의미가 있을지 모르겠네요. 또, 네이밍이 추가된 id검사 로직과는 관련없어보여요

riverandeye replied:
제가 네이밍을 적절하게 하지 못해서 의도가 잘못 전달된것 같습니다. 해당 부분을 변경하였는데, 새로 커밋한 걸 확인해주실수 있을까요? ^^7

hyokualexkwon commented on packages/server/src/user/dto/user.dto.ts:
```diff
@@ -0,0 +1,9 @@
+export class UserDto {
```
UserDto는 유저를 생성하거나 업데이트할 때 사용하는것 같은데, 차라리 UserRequestDto는 어떨까요? response와 대비되는 개념처럼말이죠

riverandeye replied:
좋은 의견이네요 수정하겠습니다 :smile:

hyokualexkwon commented on packages/server/src/user/user.service.ts:
```diff
@@ -14,21 +15,52 @@ export class UserService {
+    return this.createUserResponse(user);
+  }
+
+  public async putUser(putUserDto: UserDto): Promise<UserResponseDto> {
```
service는 db와 맞닿아 있다는 점에서 HTTP의 동사보다는 db에서 사용하는 동사를 사용하는게 어떨까요? put 보다는 update라는 네이밍은 어떨까요?

riverandeye replied:
반영하였습니다 ~.~

hyokualexkwon commented on packages/server/src/user/user.service.ts:
```diff
@@ -14,21 +15,52 @@ export class UserService {
+    });
+    this.userRepository.merge(user, putUserDto);
+    const result = await this.userRepository.save(user);
+    return this.createUserResponse(user);
```
이렇게 작성하면 업데이트 이전의 결과를 반환할것 같은데, 아닌가요?

riverandeye replied:
this.userRepository.merge 를 하게 되면, 위에 findOne으로 검색된 user의 값이 변경됩니다. 
save를 할 때 변경된 user를 save 하는 것이기 때문에 save가 성공할 시 새로 변경된 user가 올바르게 반환됩니다.

hyokualexkwon replied:
아하!

Atercatus replied:
response 객체는 controller에서 하는 것으로 통일 하는게 좋을거 같습니다.

riverandeye committed:
[server] | refactor(user): 코드 네이밍 수정

validate-id-param.pipe 가 적절하지 못하다고 생각되어
id-param.pipe 로 수정하였습니다.

user.service의 putUser를 updateUser로 변경하였습니다.
user.dto.ts 를 userRequest.dto.ts 로 변경하였습니다.

hyokualexkwon commented on packages/server/src/common/pipes/id-param.pipe/index.ts:
```diff
@@ -0,0 +1,24 @@
+  }
+
+  private validateIdParamPipeDto({ id }: IdParamPipeDto) {
+    return id && this.validateId(id);
```
id의 존재 유무를 그냥 validateId에서 하면 안되나요?

riverandeye replied:
사실 그래도 되요 ㅋㅋ 그냥 이전에 만든 pipe랑 컨벤션을 비슷하게 하려고 한건데 불필요하다 싶으시면 수정할게여

hyokualexkwon commented on packages/server/src/user/user.controller.ts:
```diff
@@ -1,18 +1,57 @@
-    return await this.userService.findVideos(userId);
+  @Get('/:id/videos')
+  @UsePipes(IdParamPipe)
+  public async getVideos(@Param() { id }: UserIdParamDto): Promise<Video[]> {
```
ㅋㅋ getUserVideos? getVideos? ㅋㅋ 뭐가 조으깡

riverandeye replied:
User 에 있는 getVideos라서 괜찮지않을까여? ㅋㅋ

Atercatus replied:
다르게 좋을 거 같네요

Atercatus replied:
`findUserVideo`는 어떤가요

hyokualexkwon commented on packages/server/src/user/user.service.ts:
```diff
@@ -14,21 +15,56 @@ export class UserService {
-      },
-    });
+  public async findUser(id: number): Promise<UserResponseDto> {
+    const user = await this.userRepository.findOne(id);
```
만약에 해당 id를 가지는 유저가 없어서 null이 반환되면, exception 처리는 안해도 될까요?

Atercatus commented on packages/server/src/common/pipes/id-param.pipe/index.ts:
```diff
@@ -0,0 +1,24 @@
+import { IdParamPipeDto } from './dto';
+
+@Injectable()
+export class IdParamPipe implements PipeTransform {
```
이름이 모호한거 같습니다. 어떤 Id 를 뜻하는지 그리고 Pipe가 하는 역할이 무엇인지 좀 더 명확하게 나타낼 수 있으면 좋겠습니다.
ex) transform, validate

Atercatus commented on packages/server/src/common/pipes/id-param.pipe/index.ts:
```diff
@@ -0,0 +1,24 @@
+    return { id: parseInt(id, 10) };
+  }
+
+  private validateIdParamPipeDto({ id }: IdParamPipeDto) {
```
위에 존재하는 `validateIdPipeDto` 변수와 이름이 많이 겹치고 이해하기가 힘든거 같습니다. 다른 이름을 지정해야 할거 같아요

Atercatus commented on packages/server/src/user/pipes/validate-user.pipe/dto.ts:
```diff
@@ -0,0 +1,9 @@
+export class ValidateUserPipeDto {
```
이건 동사일까요 명사일까요

Atercatus commented on packages/server/src/user/user.service.ts:
```diff
@@ -14,21 +15,56 @@ export class UserService {
   }
+
+  public async insertUser(
+    createUserDto: UserRequestDto,
```
함수 이름과, 매개변수 이름과 타입등이 어색합니다

Atercatus commented on packages/server/src/user/user.service.ts:
```diff
@@ -14,21 +15,56 @@ export class UserService {
+    return this.createUserResponse(user);
+  }
+
+  public async updateUser(
```
아직 사용되지 않는 서비스를 많이 만들지 말아주세요. 테스트하기가 힘들고, 리뷰하는 사람으로 하여금 많은 상상력을 요구하게 됩니다.

Atercatus replied:
기본적인 네이밍과 로직을 고정시키는 것은 좋지 못한 컨벤션인거 같습니다. 공통적인 부분에 대해서 일괄적인 네이밍과 로직 구조는 도움이 되겠지만 복붙 또는 따라하기 위한 컨벤션은 기피해야할 부분인거 같습니다.

riverandeye committed:
[server] | refactor(user): user 컨트롤러, 서비스, 파이프 리팩토링
